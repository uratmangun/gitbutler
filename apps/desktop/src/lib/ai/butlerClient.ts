import { messageParamToPrompt, splitPromptMessages } from './anthropicUtils';
import {
	SHORT_DEFAULT_BRANCH_TEMPLATE,
	SHORT_DEFAULT_COMMIT_TEMPLATE,
	SHORT_DEFAULT_PR_TEMPLATE
} from '$lib/ai/prompts';
import { ModelKind, type AIClient, type Prompt } from '$lib/ai/types';
import { map, wrapAsync, type Result } from '$lib/result';
import type { HttpClient } from '@gitbutler/shared/httpClient';

function splitPromptMessagesIfNecessary(
	modelKind: ModelKind,
	prompt: Prompt
): [Prompt, string | undefined] {
	switch (modelKind) {
		case ModelKind.Anthropic: {
			const [messages, system] = splitPromptMessages(prompt);
			return [messageParamToPrompt(messages), system];
		}
		case ModelKind.OpenAI:
		case ModelKind.Ollama:
			return [prompt, undefined];
	}
}

export class ButlerAIClient implements AIClient {
	defaultCommitTemplate = SHORT_DEFAULT_COMMIT_TEMPLATE;
	defaultBranchTemplate = SHORT_DEFAULT_BRANCH_TEMPLATE;
	defaultPRTemplate = SHORT_DEFAULT_PR_TEMPLATE;

	constructor(
		private cloud: HttpClient,
		private userToken: string,
		private modelKind: ModelKind
	) {}

	async evaluate(prompt: Prompt): Promise<Result<string, Error>> {
		const [messages, system] = splitPromptMessagesIfNecessary(this.modelKind, prompt);
		const response = await wrapAsync<{ message: string }, Error>(
			async () =>
				await this.cloud.post<{ message: string }>('evaluate_prompt/predict.json', {
					body: {
						messages,
						system,
						max_tokens: 400,
						model_kind: this.modelKind
					},
					token: this.userToken
				})
		);

		return map(response, ({ message }) => message);
	}
}
