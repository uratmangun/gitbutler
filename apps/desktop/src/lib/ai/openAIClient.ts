import {
	SHORT_DEFAULT_BRANCH_TEMPLATE,
	SHORT_DEFAULT_COMMIT_TEMPLATE,
	SHORT_DEFAULT_PR_TEMPLATE
} from '$lib/ai/prompts';
import OpenAI from 'openai';
import type { OpenAIModelName, Prompt, AIClient, AIEvalOptions } from '$lib/ai/types';

const DEFAULT_MAX_TOKENS = 1024;

export class OpenAIClient implements AIClient {
	defaultCommitTemplate = SHORT_DEFAULT_COMMIT_TEMPLATE;
	defaultBranchTemplate = SHORT_DEFAULT_BRANCH_TEMPLATE;
	defaultPRTemplate = SHORT_DEFAULT_PR_TEMPLATE;

	private client: OpenAI;
	private openAIKey: string;
	private modelName: OpenAIModelName;

	constructor(openAIKey: string, modelName: OpenAIModelName) {
		this.openAIKey = openAIKey;
		this.modelName = modelName;
		this.client = new OpenAI({ apiKey: openAIKey, dangerouslyAllowBrowser: true });
	}

	async evaluate(prompt: Prompt, options?: AIEvalOptions): Promise<string> {
		const response = await this.client.chat.completions.create({
			max_tokens: options?.maxTokens ?? DEFAULT_MAX_TOKENS,
			messages: prompt,
			model: this.modelName,
			stream: true
		});

		const buffer: string[] = [];
		for await (const chunk of response) {
			const token = chunk.choices[0]?.delta.content ?? '';
			options?.onToken?.(token);
			buffer.push(token);
		}
		return buffer.join('');
	}
}
