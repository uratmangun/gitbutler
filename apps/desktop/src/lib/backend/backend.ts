import type { Readable } from 'svelte/store';
type Event<T> = {
	/** Event name */
	event: string;
	/** Event identifier used to unlisten */
	id: number;
	/** Event payload */
	payload: T;
};

export type DownloadEvent =
	| {
			event: 'Started';
			data: {
				contentLength?: number;
			};
	  }
	| {
			event: 'Progress';
			data: {
				chunkLength: number;
			};
	  }
	| {
			event: 'Finished';
	  };

export type DownloadEventName = DownloadEvent['event'];

export type DownloadUpdate = (onEvent?: (progress: DownloadEvent) => void) => Promise<void>;
export type InstallUpdate = () => Promise<void>;

export type Update = {
	version: string;
	currentVersion: string;
	body?: string;
	download: DownloadUpdate;
	install: InstallUpdate;
};

export interface IBackend {
	systemTheme: Readable<string | null>;
	invoke: <T>(command: string, ...args: any[]) => Promise<T>;
	listen: <T>(event: string, callback: (event: Event<T>) => void) => () => Promise<void>;
	checkUpdate: () => Promise<Update | null>;
	currentVersion: () => Promise<string>;
	readFile: (path: string) => Promise<Uint8Array>;
	openExternalUrl: (href: string) => Promise<void>;
	relaunch: () => Promise<void>;
}
