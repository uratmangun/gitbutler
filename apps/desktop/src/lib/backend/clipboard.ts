import { writeText, readText } from '@tauri-apps/plugin-clipboard-manager';

export async function writeClipboard(text: string) {
	await writeText(text);
}

export async function readClipboard() {
	return await readText();
}
