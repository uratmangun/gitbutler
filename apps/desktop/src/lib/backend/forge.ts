import { invoke } from './ipc';

export type ForgeType = 'github' | 'gitlab' | 'bitbucket' | 'azure';

export class ForgeService {
	constructor(private projectId: string) {}

	async getAvailableReviewTemplates(): Promise<string[]> {
		const templates = await invoke<string[]>('get_available_review_templates', {
			projectId: this.projectId
		});

		return templates;
	}
}
