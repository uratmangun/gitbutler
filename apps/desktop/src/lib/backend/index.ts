import Tauri from '$lib/backend/tauri';
import Web from '$lib/backend/web';
import { InjectionToken } from '@gitbutler/shared/context';
import type { IBackend } from '$lib/backend/backend';

export const BACKEND = new InjectionToken<IBackend>('Backend');

export default function createBackend(): IBackend {
	if (import.meta.env.VITE_BUILD_TARGET === 'web') {
		return new Web();
	}
	return new Tauri();
}

export function isBackend(something: unknown): something is IBackend {
	return (
		typeof something === 'object' &&
		something !== null &&
		(something instanceof Tauri || something instanceof Web)
	);
}

export * from '$lib/backend/backend';
