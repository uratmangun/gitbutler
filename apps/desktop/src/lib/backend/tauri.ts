import { isReduxError } from '$lib/state/reduxError';
import { getVersion as tauriGetVersion } from '@tauri-apps/api/app';
import { invoke as invokeTauri } from '@tauri-apps/api/core';
import { listen as listenTauri } from '@tauri-apps/api/event';
import { documentDir as documentDirTauri } from '@tauri-apps/api/path';
import { join as joinPathTauri } from '@tauri-apps/api/path';
import { getCurrentWindow, Window } from '@tauri-apps/api/window';
import { open as filePickerTauri, type OpenDialogOptions } from '@tauri-apps/plugin-dialog';
import { readFile as tauriReadFile } from '@tauri-apps/plugin-fs';
import { relaunch as relaunchTauri } from '@tauri-apps/plugin-process';
import { check as tauriCheck } from '@tauri-apps/plugin-updater';
import { readable } from 'svelte/store';
import type { IBackend } from '$lib/backend/backend';
import type { EventCallback, EventName } from '@tauri-apps/api/event';

export default class Tauri implements IBackend {
	private appWindow: Window | undefined;
	systemTheme = readable<string | null>(null, (set) => {
		if (!this.appWindow) {
			this.appWindow = getCurrentWindow();
		}
		this.appWindow.theme().then((value) => {
			set(value);
		});

		this.appWindow.onThemeChanged((e) => {
			set(e.payload);
		});
	});
	invoke = tauriInvoke;
	listen = tauriListen;
	checkUpdate = tauriCheck;
	currentVersion = tauriGetVersion;
	readFile = tauriReadFile;
	openExternalUrl = tauriOpenExternalUrl;
	relaunch = relaunchTauri;
	documentDir = documentDirTauri;
	joinPath = joinPathTauri;
	async filePicker<T extends OpenDialogOptions>() {
		return await filePickerTauri<T>();
	}
	async homeDirectory(): Promise<string> {
		// TODO: Find a workaround to avoid this dynamic import
		// https://github.com/sveltejs/kit/issues/905
		return await (await import('@tauri-apps/api/path')).homeDir();
	}
}

async function tauriInvoke<T>(command: string, params: Record<string, unknown> = {}): Promise<T> {
	// This commented out code can be used to delay/reject an api call
	// return new Promise<T>((resolve, reject) => {
	// 	if (command.startsWith('apply')) {
	// 		setTimeout(() => {
	// 			reject('testing the error page');
	// 		}, 500);
	// 	} else {
	// 		resolve(invokeTauri<T>(command, params));
	// 	}
	// }).catch((reason) => {
	// 	const userError = UserError.fromError(reason);
	// 	console.error(`ipc->${command}: ${JSON.stringify(params)}`, userError);
	// 	throw userError;
	// });

	try {
		return await invokeTauri<T>(command, params);
	} catch (error: unknown) {
		if (isReduxError(error)) {
			console.error(`ipc->${command}: ${JSON.stringify(params)}`, error);
		}
		throw error;
	}
}

function tauriListen<T>(event: EventName, handle: EventCallback<T>) {
	const unlisten = listenTauri(event, handle);
	return async () => await unlisten.then((unlistenFn) => unlistenFn());
}

async function tauriOpenExternalUrl(href: string): Promise<void> {
	return await invokeTauri<void>('open_url', { url: href });
}
