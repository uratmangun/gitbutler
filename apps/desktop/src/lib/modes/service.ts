import { invoke, listen } from '$lib/backend/ipc';
import { derived, writable } from 'svelte/store';

type Mode = { type: 'OpenWorkspace' } | { type: 'OutsideWorksapce' } | { type: 'Edit' };

export class ModeService {
	readonly head = writable<string | undefined>(undefined);
	readonly mode = writable<Mode | undefined>(undefined);

	readonly gbBranchActive = derived(this.head, (head) => head === 'gitbutler/integration');

	unsubscribe?: () => Promise<void>;

	constructor(private projectId: string) {
		this.unsubscribe = subscribeToHead(projectId, (head, mode) => {
			this.head.set(head);
			this.mode.set(mode);
		});
		this.refresh();
	}

	private async refresh() {
		const head = await invoke<string>('git_head', { projectId: this.projectId });
		this.head.set(head);

		const mode = await invoke<Mode>('operating_mode', { projectId: this.projectId });
		this.mode.set(mode);
	}
}

function subscribeToHead(projectId: string, callback: (head: string, mode: Mode) => void) {
	return listen<{ head: string; operating_mode: Mode }>(
		`project://${projectId}/git/head`,
		(event) => callback(event.payload.head, event.payload.operating_mode)
	);
}
