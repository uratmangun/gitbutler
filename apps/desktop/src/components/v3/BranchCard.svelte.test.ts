import BranchCard from '$components/v3/BranchCard.svelte';
import { AIService } from '$lib/ai/service';
import { Tauri } from '$lib/backend/tauri';
import { BaseBranch } from '$lib/baseBranch/baseBranch';
import BaseBranchService from '$lib/baseBranch/baseBranchService.svelte';
import { DefaultForgeFactory } from '$lib/forge/forgeFactory.svelte';
import { StackService } from '$lib/stacks/stackService.svelte';
import { UiState } from '$lib/state/uiState.svelte';
import { getAIServiceMock } from '$lib/testing/mockAIService';
import { getMockBaseBranch } from '$lib/testing/mockBaseBranch';
import { getStackServiceMock } from '$lib/testing/mockStackService';
import { getUiStateMock } from '$lib/testing/mockUiState';
import { setup } from '$lib/testing/setup';
import { TestId } from '$lib/testing/testIds';
import { clearMocks } from '@tauri-apps/api/mocks';
import { render, waitFor } from '@testing-library/svelte/svelte5';
import userEvent from '@testing-library/user-event';
import { expect, test, describe, beforeEach, afterEach, vi } from 'vitest';

describe('BranchCard Component', () => {
	let cleanup: () => void;

	let tauri: Tauri;
	let stackService: ReturnType<typeof vi.fn>;
	let baseBranchService: ReturnType<typeof vi.fn>;
	let uiState: ReturnType<typeof vi.fn>;
	let forge: ReturnType<typeof vi.fn>;
	let aiService: ReturnType<typeof vi.fn>;
	let baseBranch: ReturnType<typeof vi.fn>;

	let context: Map<any, any>;

	beforeEach(() => {
		cleanup = setup();

		tauri = new Tauri();
		const MockStackService = getStackServiceMock();
		const UiStateMock = getUiStateMock();
		const BaseBranchMock = getMockBaseBranch();
		const AIServiceMock = getAIServiceMock();

		stackService = new MockStackService();
		baseBranchService = vi.fn();
		uiState = new UiStateMock();
		forge = vi.fn();
		aiService = new AIServiceMock();
		baseBranch = new BaseBranchMock();

		context = new Map<any, any>([
			[StackService, stackService],
			[BaseBranchService, baseBranchService],
			[UiState, uiState],
			[DefaultForgeFactory, forge],
			[AIService, aiService],
			[BaseBranch, baseBranch]
		]);
		vi.spyOn(tauri, 'listen').mockReturnValue(async () => {});
	});

	afterEach(() => {
		vi.restoreAllMocks();
		clearMocks();

		cleanup();
	});

	test('should render the BranchCard component and update it correctly', async () => {
		const StackServiceMock = getStackServiceMock();

		// Override the stack service in the context to inject spies 🕵🏻‍♂️
		const branchByNameMock = vi.spyOn(StackServiceMock.prototype, 'branchByName');
		const branchesMock = vi.spyOn(StackServiceMock.prototype, 'branches');
		const branchDetailsMock = vi.spyOn(StackServiceMock.prototype, 'branchDetails');
		const commitAtMock = vi.spyOn(StackServiceMock.prototype, 'commitAt');
		const commitsMock = vi.spyOn(StackServiceMock.prototype, 'commits');
		const upstreamCommitsMock = vi.spyOn(StackServiceMock.prototype, 'upstreamCommits');

		const stackServiceMock = new StackServiceMock();
		context.set(StackService, stackServiceMock);

		const branchName = 'branch-a';
		const { getByTestId, rerender } = render(BranchCard, {
			context,
			props: {
				projectId: 'test-project',
				stackId: 'test-stack',
				branchName,
				first: true,
				last: true
			}
		});

		await waitFor(
			() => {
				const branchLabel = getByTestId(TestId.BranchNameLabel);
				expect(branchLabel).toBeInTheDocument();
				expect(branchLabel).toHaveTextContent(branchName);
			},
			{ timeout: 5000 }
		);

		expect(branchByNameMock).toHaveBeenCalledWith('test-project', 'test-stack', branchName);
		expect(branchByNameMock).toHaveBeenCalledTimes(1);

		expect(branchesMock).toHaveBeenCalledWith('test-project', 'test-stack');
		expect(branchesMock).toHaveBeenCalledTimes(1);

		expect(branchDetailsMock).toHaveBeenCalledWith('test-project', 'test-stack', branchName);
		expect(branchDetailsMock).toHaveBeenCalledTimes(1);

		expect(commitAtMock).toHaveBeenCalledWith('test-project', 'test-stack', branchName, 0);

		expect(commitsMock).toHaveBeenCalledWith('test-project', 'test-stack', branchName);

		expect(upstreamCommitsMock).toHaveBeenCalledWith('test-project', 'test-stack', branchName);

		// Update the branch name props should trigger all stack service calls again

		await rerender({
			projectId: 'test-project',
			stackId: 'test-stack',
			branchName: 'branch-b',
			first: true,
			last: true
		});

		expect(branchByNameMock).toHaveBeenCalledWith('test-project', 'test-stack', 'branch-b');
		expect(branchByNameMock).toHaveBeenCalledTimes(2);

		expect(branchesMock).toHaveBeenCalledWith('test-project', 'test-stack');
		// Since none of the branches parameters changed, the branches call should not be triggered again
		expect(branchesMock).toHaveBeenCalledTimes(1);

		expect(branchDetailsMock).toHaveBeenCalledWith('test-project', 'test-stack', 'branch-b');
		expect(branchDetailsMock).toHaveBeenCalledTimes(2);

		expect(commitAtMock).toHaveBeenCalledWith('test-project', 'test-stack', 'branch-b', 0);
		expect(commitAtMock.mock.calls).toMatchSnapshot();

		expect(commitsMock).toHaveBeenCalledWith('test-project', 'test-stack', 'branch-b');
		expect(commitsMock.mock.calls).toMatchSnapshot();

		expect(upstreamCommitsMock).toHaveBeenCalledWith('test-project', 'test-stack', 'branch-b');
		expect(upstreamCommitsMock.mock.calls).toMatchSnapshot();
	});

	test('should display the context menu on right click', async () => {
		const user = userEvent.setup();

		const { getByTestId, queryByTestId } = render(BranchCard, {
			context,
			props: {
				projectId: 'test-project',
				stackId: 'test-stack',
				branchName: 'branch-a',
				first: true,
				last: true
			}
		});

		const branchHeader = getByTestId(TestId.BranchHeader);
		expect(branchHeader).toBeInTheDocument();

		await user.pointer({ keys: '[MouseRight]', target: branchHeader });

		const contextMenu = getByTestId(TestId.BranchHeaderContextMenu);
		expect(contextMenu).toBeInTheDocument();
		expect(contextMenu).toBeVisible();

		// Verify context menu items
		const addDependentBranchItem = getByTestId(TestId.BranchHeaderContextMenu_AddDependentBranch);
		expect(addDependentBranchItem).toBeInTheDocument();

		// Should not be able to open in browser
		const openInBrowserItem = queryByTestId(TestId.BranchHeaderContextMenu_OpenInBrowser);
		expect(openInBrowserItem).toBeNull();

		const copyBranchNameItem = getByTestId(TestId.BranchHeaderContextMenu_CopyBranchName);
		expect(copyBranchNameItem).toBeInTheDocument();

		// Should not be able to add/remove description
		const addRemoveDescriptionItem = queryByTestId(
			TestId.BranchHeaderContextMenu_AddRemoveDescription
		);
		expect(addRemoveDescriptionItem).toBeNull();

		// Should not be able to generate branch name
		const generateBranchNameItem = queryByTestId(TestId.BranchHeaderContextMenu_GenerateBranchName);
		expect(generateBranchNameItem).toBeNull();

		const renameItem = getByTestId(TestId.BranchHeaderContextMenu_Rename);
		expect(renameItem).toBeInTheDocument();

		// Should not be able to delete, since it's the only branch
		const deleteItem = queryByTestId(TestId.BranchHeaderContextMenu_Delete);
		expect(deleteItem).toBeNull();

		// Should not be able to open PR in browser, sice there's no PR
		const openPRInBrowserItem = queryByTestId(TestId.BranchHeaderContextMenu_OpenPRInBrowser);
		expect(openPRInBrowserItem).toBeNull();

		// Should not be able to copy PR link, since there's no PR
		const copyPRLinkItem = queryByTestId(TestId.BranchHeaderContextMenu_CopyPRLink);
		expect(copyPRLinkItem).toBeNull();
	});
});
