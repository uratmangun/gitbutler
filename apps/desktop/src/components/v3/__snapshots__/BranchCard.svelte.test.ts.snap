// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BranchCard Component > should render the BranchCard component and update it correctly 1`] = `
[
  [
    "test-project",
    "test-stack",
    "branch-a",
    0,
  ],
  [
    "test-project",
    "test-stack",
    "branch-a",
    0,
  ],
  [
    "test-project",
    "test-stack",
    "branch-b",
    0,
  ],
  [
    "test-project",
    "test-stack",
    "branch-b",
    0,
  ],
]
`;

exports[`BranchCard Component > should render the BranchCard component and update it correctly 2`] = `
[
  [
    "test-project",
    "test-stack",
    "branch-a",
  ],
  [
    "test-project",
    "test-stack",
    "branch-a",
  ],
  [
    "test-project",
    "test-stack",
    "branch-b",
  ],
  [
    "test-project",
    "test-stack",
    "branch-b",
  ],
]
`;

exports[`BranchCard Component > should render the BranchCard component and update it correctly 3`] = `
[
  [
    "test-project",
    "test-stack",
    "branch-a",
  ],
  [
    "test-project",
    "test-stack",
    "branch-a",
  ],
  [
    "test-project",
    "test-stack",
    "branch-b",
  ],
  [
    "test-project",
    "test-stack",
    "branch-b",
  ],
]
`;
